<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Framework for the analysis of genomic prediction data using R</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Framework for the analysis of genomic prediction data using R
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;synbreed&rsquo; version 0.10-2</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../demo">Code demos</a>.  Use <a href="../../utils/help/demo">demo()</a> to run them.</li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="add.individuals.html">add.individuals</a></td>
<td>Add new individuals to objects of class gpData</td></tr>
<tr><td width="25%"><a href="add.markers.html">add.markers</a></td>
<td>Add new markers to an object of class gpData</td></tr>
<tr><td width="25%"><a href="codeGeno.html">codeGeno</a></td>
<td>Recode genotypic data, imputation of missing values and preselection of markers</td></tr>
<tr><td width="25%"><a href="create.gpData.html">create.gpData</a></td>
<td>Create genomic prediction data object</td></tr>
<tr><td width="25%"><a href="create.pedigree.html">create.pedigree</a></td>
<td>Create pedigree object</td></tr>
<tr><td width="25%"><a href="gpData2cross.html">cross2gpData</a></td>
<td>Conversion between objects of class 'cross' and 'gpData'</td></tr>
<tr><td width="25%"><a href="crossVal.html">crossVal</a></td>
<td>Cross validation of different prediction models</td></tr>
<tr><td width="25%"><a href="discard.markers.html">discard.individuals</a></td>
<td>Subsets for objects of class gpData</td></tr>
<tr><td width="25%"><a href="discard.markers.html">discard.markers</a></td>
<td>Subsets for objects of class gpData</td></tr>
<tr><td width="25%"><a href="gpData2cross.html">gpData2cross</a></td>
<td>Conversion between objects of class 'cross' and 'gpData'</td></tr>
<tr><td width="25%"><a href="gpData2data.frame.html">gpData2data.frame</a></td>
<td>Merge of phenotypic and genotypic data</td></tr>
<tr><td width="25%"><a href="gpMod.html">gpMod</a></td>
<td>Genomic predictions models for objects of class gpData</td></tr>
<tr><td width="25%"><a href="kinship.html">kin</a></td>
<td>Relatedness based on pedigree or marker data</td></tr>
<tr><td width="25%"><a href="LDDist.html">LDDist</a></td>
<td>LD versus distance Plot</td></tr>
<tr><td width="25%"><a href="LDMap.html">LDMap</a></td>
<td>LD Heatmap</td></tr>
<tr><td width="25%"><a href="manhattanPlot.html">manhattanPlot</a></td>
<td>Manhattan plot for SNP effects</td></tr>
<tr><td width="25%"><a href="MME.html">MME</a></td>
<td>Mixed Model Equations</td></tr>
<tr><td width="25%"><a href="LD.html">pairwiseLD</a></td>
<td>Pairwise LD between markers</td></tr>
<tr><td width="25%"><a href="plot.genMap.html">plot.GenMap</a></td>
<td>Plot marker map</td></tr>
<tr><td width="25%"><a href="plot.LDdf.html">plot.LDdf</a></td>
<td>Plot function for class 'LDdf'</td></tr>
<tr><td width="25%"><a href="plot.LDmat.html">plot.LDmat</a></td>
<td>Plot function for class 'LDmat'</td></tr>
<tr><td width="25%"><a href="plot.pedigree.html">plot.pedigree</a></td>
<td>Visualization of pedigree</td></tr>
<tr><td width="25%"><a href="plot.relationshipMatrix.html">plot.relationshipMatrix</a></td>
<td>Heatmap for relationship Matrix</td></tr>
<tr><td width="25%"><a href="plot.genMap.html">plotGenMap</a></td>
<td>Plot marker map</td></tr>
<tr><td width="25%"><a href="plotNeighbourLD.html">plotNeighbourLD</a></td>
<td>Plot neighbour linkage disequilibrium</td></tr>
<tr><td width="25%"><a href="predict.gpMod.html">predict.gpMod</a></td>
<td>Prediction for genomic prediction models.</td></tr>
<tr><td width="25%"><a href="summary.cvData.html">print.summary.cvData</a></td>
<td>Summary of options and results of the cross validation procedure</td></tr>
<tr><td width="25%"><a href="summary.gpData.html">print.summary.gpData</a></td>
<td>Summary for class gpData</td></tr>
<tr><td width="25%"><a href="summary.gpMod.html">print.summary.gpMod</a></td>
<td>Summary for class gpMod</td></tr>
<tr><td width="25%"><a href="summary.gpMod.html">print.summary.gpModList</a></td>
<td>Summary for class gpMod</td></tr>
<tr><td width="25%"><a href="summary.pedigree.html">print.summary.pedigree</a></td>
<td>Summary of pedigree information</td></tr>
<tr><td width="25%"><a href="summary.relationshipMatrix.html">print.summary.relationshipMatrix</a></td>
<td>Summary of relationship matrices</td></tr>
<tr><td width="25%"><a href="simul.pedigree.html">simul.pedigree</a></td>
<td>Simulation of pedigree structure</td></tr>
<tr><td width="25%"><a href="simul.phenotype.html">simul.phenotype</a></td>
<td>Simulation of a field trial with single trait</td></tr>
<tr><td width="25%"><a href="summary.cvData.html">summary.cvData</a></td>
<td>Summary of options and results of the cross validation procedure</td></tr>
<tr><td width="25%"><a href="summary.gpData.html">summary.gpData</a></td>
<td>Summary for class gpData</td></tr>
<tr><td width="25%"><a href="summary.gpMod.html">summary.gpMod</a></td>
<td>Summary for class gpMod</td></tr>
<tr><td width="25%"><a href="summary.gpMod.html">summary.gpModList</a></td>
<td>Summary for class gpMod</td></tr>
<tr><td width="25%"><a href="summary.LDdf.html">summary.LDdf</a></td>
<td>Summary for LD objects</td></tr>
<tr><td width="25%"><a href="summary.LDdf.html">summary.LDmat</a></td>
<td>Summary for LD objects</td></tr>
<tr><td width="25%"><a href="summary.pedigree.html">summary.pedigree</a></td>
<td>Summary of pedigree information</td></tr>
<tr><td width="25%"><a href="summary.relationshipMatrix.html">summary.relationshipMatrix</a></td>
<td>Summary of relationship matrices</td></tr>
<tr><td width="25%"><a href="summary.map.html">summaryGenMap</a></td>
<td>Summary of marker map information</td></tr>
<tr><td width="25%"><a href="write.beagle.html">write.beagle</a></td>
<td>Prepare genotypic data for Beagle</td></tr>
<tr><td width="25%"><a href="write.plink.html">write.plink</a></td>
<td>Prepare data for PLINK</td></tr>
<tr><td width="25%"><a href="write.relationshipMatrix.html">write.relationshipMatrix</a></td>
<td>Writing relationshipMatrix in table format</td></tr>
<tr><td width="25%"><a href="Extract.GenMap.html">[.GenMap</a></td>
<td>Extract or replace part of map data.frame</td></tr>
<tr><td width="25%"><a href="Extract.relationshipMatrix.html">[.relationshipMatrix</a></td>
<td>Extract or replace part of relationship matrix</td></tr>
</table>
</body></html>
